src/main/java/org/terasology/Lakes/LakeProvider.java

```diff
@@ -15,182 +15,288 @@
  */
 package org.terasology.Lakes;
 
-import org.terasology.core.world.CoreBiome;
-import org.terasology.core.world.generator.facets.BiomeFacet;
-import org.terasology.math.Region3i;
-import org.terasology.math.geom.Rect2i;
-import org.terasology.math.geom.Vector2i;
+import org.terasology.math.JomlUtil;
+import org.terasology.math.geom.Vector3f;
 import org.terasology.math.geom.Vector3i;
-import org.terasology.naming.Name;
+import org.terasology.registry.CoreRegistry;
+import org.terasology.utilities.procedural.BrownianNoise;
 import org.terasology.utilities.procedural.Noise;
+import org.terasology.utilities.procedural.SimplexNoise;
+import org.terasology.utilities.procedural.SubSampledNoise;
 import org.terasology.utilities.procedural.WhiteNoise;
+import org.terasology.world.block.Block;
+import org.terasology.world.block.BlockManager;
 import org.terasology.world.generation.Border3D;
 import org.terasology.world.generation.Facet;
 import org.terasology.world.generation.FacetBorder;
 import org.terasology.world.generation.FacetProviderPlugin;
 import org.terasology.world.generation.GeneratingRegion;
 import org.terasology.world.generation.Produces;
 import org.terasology.world.generation.Requires;
-import org.terasology.world.generation.facets.SurfaceHeightFacet;
-import org.terasology.world.generation.facets.base.BaseFieldFacet2D;
+import org.terasology.world.generation.Updates;
+import org.terasology.world.generation.facets.DensityFacet;
+import org.terasology.world.generation.facets.ElevationFacet;
+import org.terasology.world.generation.facets.SeaLevelFacet;
+import org.terasology.world.generation.facets.SurfacesFacet;
 import org.terasology.world.generator.plugin.RegisterPlugin;
 
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.Queue;
+import java.util.Set;
+
+
+/**
+ * Try to place lakes in random locations on the surface and underground.
+ * The depth of each lake is determined by a random noise field added to a random-width paraboloid.
+ * The points where the depth is <= 0 are where the lake stops.
+ * If the surface in that region is missing, or too steep, or under various other conditions,
+ * the provider gives up on generating that lake, and doesn't place it.
+ * 
+ * Underground lakes are generated similarly, but in 3D.
+ * If they breach the surface, they're cancelled.
+ * The upper part of their cave is filled with air.
+ * If they're sufficiently far underground, they're lava instead of water.
+ */
 @RegisterPlugin
 @Produces(LakeFacet.class)
+@Updates({
+        @Facet(value = SurfacesFacet.class, border = @FacetBorder(sides = Lake.MAX_RADIUS * 3, bottom = Lake.MAX_DEPTH + 3, top = Lake.MAX_DEPTH + 3)),
```

Comment:
A facet that is being updated is automatically also part of `@Requires(...)`?

Reply:
Yes, it's a bit like a combination of `@Requires` and `@Produces`, although I think there's a slight problem with borders and `@Updates`.